server/block/hopper.go

```diff
@@ -0,0 +1,265 @@
+	TransferCooldown int64
+	// CollectCooldown is the duration until the hopper can collect items again.
+	CollectCooldown int64
+
+	inventory *inventory.Inventory
+	viewerMu  *sync.RWMutex
+	viewers   map[ContainerViewer]struct{}
+}
+
+// NewHopper creates a new initialised hopper. The inventory is properly initialised.
+func NewHopper() Hopper {
+	m := new(sync.RWMutex)
+	v := make(map[ContainerViewer]struct{}, 1)
+	return Hopper{
+		inventory: inventory.New(5, func(slot int, _, item item.Stack) {
+			m.RLock()
+			defer m.RUnlock()
+			for viewer := range v {
+				viewer.ViewSlotChange(slot, item)
+			}
+		}),
+		viewerMu: m,
+		viewers:  v,
+	}
+}
+
+// Model ...
+func (Hopper) Model() world.BlockModel {
+	return model.Hopper{}
+}
+
+// SideClosed ...
+func (Hopper) SideClosed(cube.Pos, cube.Pos, *world.World) bool {
+	return false
+}
+
+// BreakInfo ...
+func (h Hopper) BreakInfo() BreakInfo {
+	return newBreakInfo(3, pickaxeHarvestable, pickaxeEffective, oneOf(h)).withBlastResistance(24).withBreakHandler(func(pos cube.Pos, w *world.World, u item.User) {
+		for _, i := range h.Inventory(w, pos).Clear() {
+			dropItem(w, i, pos.Vec3())
+		}
+	})
+}
+
+// Inventory returns the inventory of the hopper.
+func (h Hopper) Inventory(w *world.World, pos cube.Pos) *inventory.Inventory {
+	return h.inventory
+}
+
+// WithName returns the hopper after applying a specific name to the block.
+func (h Hopper) WithName(a ...any) world.Item {
+	h.CustomName = strings.TrimSuffix(fmt.Sprintln(a...), "\n")
+	return h
+}
+
+// AddViewer adds a viewer to the hopper, so that it is updated whenever the inventory of the hopper is changed.
+func (h Hopper) AddViewer(v ContainerViewer, _ *world.World, _ cube.Pos) {
+	h.viewerMu.Lock()
+	defer h.viewerMu.Unlock()
+	h.viewers[v] = struct{}{}
+}
+
+// RemoveViewer removes a viewer from the hopper, so that slot updates in the inventory are no longer sent to it.
+func (h Hopper) RemoveViewer(v ContainerViewer, _ *world.World, _ cube.Pos) {
+	h.viewerMu.Lock()
+	defer h.viewerMu.Unlock()
+	delete(h.viewers, v)
+}
+
+// Activate ...
+func (Hopper) Activate(pos cube.Pos, _ cube.Face, _ *world.World, u item.User, _ *item.UseContext) bool {
+	if opener, ok := u.(ContainerOpener); ok {
+		opener.OpenBlockContainer(pos)
+		return true
+	}
+	return false
+}
+
+// UseOnBlock ...
+func (h Hopper) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) bool {
+	pos, _, used := firstReplaceable(w, pos, face, h)
+	if !used {
+		return false
+	}
+
+	//noinspection GoAssignmentToReceiver
+	h = NewHopper()
+	h.Facing = cube.FaceDown
+	if h.Facing != face {
+		h.Facing = face.Opposite()
+	}
+
+	place(w, pos, h, user, ctx)
+	return placed(ctx)
+}
+
+// Tick ...
+func (h Hopper) Tick(currentTick int64, pos cube.Pos, w *world.World) {
+	h.TransferCooldown--
+	h.CollectCooldown--
+	h.LastTick = currentTick
+
+	if !h.Powered && h.TransferCooldown < 0 {
```

Comment:
```suggestion
	if !h.Powered && h.TransferCooldown <= 0 {
```
according to Java Edition